src/onnx/parse_gridsample.cpp

```diff
@@ -0,0 +1,432 @@
+        auto h_clamp = validate ? info.add_common_op("clip", h, m_zero_l, m_height_max_l) : h;
+        auto nchw    = info.add_instruction(make_op("concat", {{"axis", 0}}), nc, h_clamp, w_clamp);
+        indices.push_back(nchw);
+        if(validate)
+        {
+            auto h_valid = info.add_common_op("equal", h, h_clamp);
+            auto w_valid = info.add_common_op("equal", w, w_clamp);
+            auto valid   = info.add_common_op("logical_and", h_valid, w_valid);
+            validation.push_back(valid);
+        }
+    }
+
+    static instruction_ref concat_on_first_dim(const onnx_parser::node_info& info,
+                                               std::vector<instruction_ref> instructions)
+    {
+        if(instructions.empty())
+            MIGRAPHX_THROW("PARSE_GRID_SAMPLE: can't concatenate empty list of instructions");
+
+        auto ret = instructions.at(0);
+        std::for_each(
+            std::next(instructions.begin()), instructions.end(), [&info, &ret](auto& ins) {
+                ret = info.add_instruction(make_op("concat", {{"axis", 0}}), ret, ins);
+            });
+        return ret;
+    }
+
+    inline bool has_border_padding() const { return m_padding == "border"; }
+
+    virtual instruction_ref sample(const onnx_parser::node_info& info) = 0;
+};
+
+struct nearest_sampler : grid_sampler
+{
+    instruction_ref m_round_x;
+    instruction_ref m_round_y;
+
+    nearest_sampler(const instruction_ref& input,
+                    const instruction_ref& grid,
+                    bool align,
+                    std::string&& padding)
+        : grid_sampler(input, grid, align, std::move(padding))
+    {
+    }
+
+    void setup(const onnx_parser::node_info& info) override
+    {
+        grid_sampler::setup(info);
+        m_round_x = info.add_common_op("nearbyint", m_unnorm_x);
+        m_round_y = info.add_common_op("nearbyint", m_unnorm_y);
+    }
+
+    instruction_ref sample(const onnx_parser::node_info& info) override
+    {
+        setup(info);
+        std::vector<instruction_ref> indices;
+        std::vector<instruction_ref> validation;
+        static auto nhw_shape = migraphx::shape{migraphx::shape::int64_type, {3}};
+        bool validate         = not has_border_padding();
+        for(size_t n = 0; n < m_batch; n++)
+        {
+            for(size_t h = 0; h < m_out_height; h++)
+            {
+                for(size_t w = 0; w < m_out_width; w++)
+                {
+                    auto nhw = info.add_literal(migraphx::literal{nhw_shape, {n, h, w}});
+                    auto h_t = info.add_instruction(make_op("gathernd"), m_round_y, nhw);
+                    auto w_t = info.add_instruction(make_op("gathernd"), m_round_x, nhw);
+                    for(size_t c = 0; c < m_channel; c++)
+                    {
+                        update_indices(info, h_t, w_t, n, c, indices, validation, validate);
+                    }
+                }
+            }
+        }
```

Comment:
I think this can be rewritten and run in parallel here per sample for each batch + height combination instead of done in serial.

Reply:
@TedThemistokleous could you please elaborate more on this. I tried `par_dfor`  to run in parellel for batch+height combination as you suggested, but it fails to compile. From what I can see `add_instruction` calls are not suitable to be executed in parallel, it produces invalid instructions because of the race condition.

Reply:
As I can see `par_dfor` is used in runtime constructions but this is a compile time problem. Is there any solution for compile time parallelism?

Reply:
I dont think par_dfor can be used, but `dfor` can be used to get rid of some of the nesting.